ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM/QD.hs

```diff
@@ -0,0 +1,875 @@
+         <> (wrapPositiveAction $ TimePasses 61)
+         <> (wrapPositiveAction $ ExtendSelection (S (-4)))
+         <> (wrapPositiveAction $ ReadMarker)
+
+prop_sequential_iosim :: Maybe UpstreamPeer -> LedgerStateJudgement -> QC.Fun (Set.Set UpstreamPeer) Bool -> QC.Property
+prop_sequential_iosim pUpstreamPeerBound pInitialJudgement (QC.Fun _ pIsHaaSatisfied) =
+    Reflection.reify (StaticParams{ pUpstreamPeerBound
+                                  , pInitialJudgement
+                                  , pIsHaaSatisfied
+                                  }) $ \(Proxy :: Proxy params) ->
+     -- NOTE: the actions have to be generated with an explicit call, as
+     -- 'generator' relies on reflection for access to the parameters
+     QC.forAll QC.arbitrary $ \actions ->
+       runIOSimProp $ prop_sequential_iosim1 @_ @params actions
+
+setupGsm :: (Set.Set UpstreamPeer -> Bool) -> SystemStateVars (IOSim.IOSim s) -> GSM.GsmEntryPoints (IOSim.IOSim s)
+setupGsm isHaaSatisfied vars = do
+  let tracer = Tracer $ push varEvents . EvGsm
+  GSM.realGsmEntryPoints (id, tracer) GSM.GsmView {
+            GSM.antiThunderingHerd = Nothing
+          ,
+            GSM.candidateOverSelection = \ s (PeerState c _) -> candidateOverSelection s c
+          ,
+            GSM.peerIsIdle = isIdling
+          ,
+            GSM.durationUntilTooOld = Just durationUntilTooOld
+          ,
+            GSM.equivalent = (==)   -- unsound, but harmless in this test
+          ,
+            GSM.getChainSyncStates = readTVar varStates
+          ,
+            GSM.getCurrentSelection = readTVar varSelection
+          ,
+            GSM.minCaughtUpDuration = thrashLimit
+          ,
+            GSM.setCaughtUpPersistentMark = \b ->
+                atomically $ do
+                    writeTVar varMarker $ if b then Present else Absent
+          ,
+            GSM.writeGsmState = \x -> atomically $ do
+                writeTVar varGsmState x
+          ,
+            GSM.isHaaSatisfied =
+                isHaaSatisfied . Map.keysSet <$> readTVar varStates
+          }
+  where
+    SystemStateVars{varSelection, varStates, varGsmState, varMarker, varEvents} = vars
+
+prop_sequential_iosim1 ::
+  forall s params.
+  Reflection.Reifies params StaticParams =>
+  QD.Actions (Model params) ->
+  QC.PropertyM (RunMonad (IOSim.IOSim s)) QC.Property
+prop_sequential_iosim1 actions = do
+  vars@SystemStateVars{varEvents} <- lift ask
+
+  let StaticParams{pInitialJudgement, pIsHaaSatisfied} = Reflection.reflect (Proxy @params)
+
+  -- initialise the SUT state using the model params
+  lift . lift $ initVars pInitialJudgement vars
+
+  let model = QD.initialState
+  let gsm = setupGsm pIsHaaSatisfied vars
+      gsmEntryPoint = case pInitialJudgement of
+            TooOld      -> GSM.enterPreSyncing gsm
+            YoungEnough -> GSM.enterCaughtUp   gsm
+
+  lift . lift $ yieldSeveralTimes
+
+  -- TODO: figure out how to do withAsync here
```

Comment:
I think what you have here is fine for now. Implementing `MonadAsync` for `PropertyM` seems dubious as `PropertyM` is `ContT` internally, and even `MonadBaseControl` doesn't allow to lift over `ContT`.

Looking at the source of `QD.runActions`, it seems that it mostly just calls [`monitor`](https://hackage.haskell.org/package/QuickCheck-2.15.0.1/docs/Test-QuickCheck-Monadic.html#v:monitor) for labelling/counterexample output, as well as asserting some pre- and postconditions. Therefore, maybe it is possible to have sth like
```haskell
runActions' :: Actions state -> m (Annotated state, Env, Property -> Property)
```
(of which `runActions` is just a small wrapper). Then, using `withAsync` should become trivial. But I don't think this has to block this PR.

Reply:
I'll leave the calls to `async` and `uninterruptibleCancel` as-is for this PR, thanks Alex!